File: lib/domain/field_cultivation/dtos/field_cultivation_sync_allocation_input.rb
# frozen_string_literal: true

module Domain
  module FieldCultivation
    module Dtos
      # 計画単位同期で扱う 1 allocation 行分（agrr 非依存）。
      class FieldCultivationSyncAllocationInput
        attr_reader :allocation_id,
                    :external_allocation_id,
                    :crop_id,
                    :start_date,
                    :completion_date,
                    :area_used,
                    :area,
                    :total_cost,
                    :cost,
                    :expected_revenue,
                    :revenue,
                    :profit,
                    :accumulated_gdd

        def initialize(
          allocation_id: nil,
          external_allocation_id: nil,
          crop_id:,
          start_date:,
          completion_date:,
          area_used: nil,
          area: nil,
          total_cost: nil,
          cost: nil,
          expected_revenue: nil,
          revenue: nil,
          profit: nil,
          accumulated_gdd: nil
        )
          @allocation_id = allocation_id
          @external_allocation_id = external_allocation_id
          @crop_id = crop_id
          @start_date = start_date
          @completion_date = completion_date
          @area_used = area_used
          @area = area
          @total_cost = total_cost
          @cost = cost
          @expected_revenue = expected_revenue
          @revenue = revenue
          @profit = profit
          @accumulated_gdd = accumulated_gdd
        end

        # duplicate 検査や FieldCultivation id 解決に使う生の識別子
        # @return [Object, nil]
        def resolved_allocation_raw
          allocation_id.nil? ? external_allocation_id : allocation_id
        end
      end
    end
  end
end

File: crates/agrr-domain/src/field_cultivation/dtos/field_cultivation_sync_allocation_input.rs
use time::Date;

#[derive(Debug, Clone, PartialEq)]
pub struct FieldCultivationSyncAllocationInput {
    pub allocation_id: Option<i64>,
    pub external_allocation_id: Option<String>,
    pub crop_id: String,
    pub start_date: Date,
    pub completion_date: Date,
    pub area_used: Option<f64>,
    pub area: Option<f64>,
    pub total_cost: Option<f64>,
    pub cost: Option<f64>,
    pub expected_revenue: Option<f64>,
    pub revenue: Option<f64>,
    pub profit: Option<f64>,
    pub accumulated_gdd: Option<f64>,
}

impl FieldCultivationSyncAllocationInput {
    pub fn resolved_allocation_raw(&self) -> Option<String> {
        if let Some(id) = self.allocation_id {
            Some(id.to_string())
        } else {
            self.external_allocation_id.clone()
        }
    }
}
